### PR TITLE
Implement functionality

### PR DESCRIPTION
**Pull request by ArjenMiedema**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Magento 2 Alternate URLs
+
+Makes it possible to configure and add alternate URLs to the HEAD of each page.
```
Comment (janmartenjongerius) [inline README.md]: Can you replace instances of "HEAD" with `<head>` ? This reads confusing in the sense that it reminds of HTTP HEAD requests.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+"connect" websites/stores to each other. The script will automatically add 
+references to the HEAD of the page for each link that is found for the current
+page.
+
+## Extending for other page types
+
+To extend this functionality to other page types (like blog posts from a module),
+you will have to create a custom module and add a custom type that implements 
+`Elgentos\AlternateUrls\Type\TypeInterface`. After that you'll need to add your
+custom type to the `typeInstances` in the `alternate_urls` block in `default.xml`.
```
Comment (janmartenjongerius) [inline README.md:28]: Can you add a small example XML block below this paragraph?

File: phpunit.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" bootstrap="vendor/autoload.php" cacheResult="false">
```
Comment (janmartenjongerius) [inline phpunit.xml]: Can you use the XSD path relative to vendor? So `vendor/phpunit/phpunit/phpunit.xsd`

That way you always have an XSD that is relevant to the installed version of PHPUnit.

File: src/Block/Adminhtml/Form/Field/AlternateUrls.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        $this->_addButtonLabel = __('Add Row')->getText();
+    }
+
+    /**
+     * @return BlockInterface
+     * @throws LocalizedException
+     */
+    private function getStoreCodeRenderer(): BlockInterface
+    {
+        if (!$this->storeCodeRenderer) {
```
Comment (janmartenjongerius) [inline src/Block/Adminhtml/Form/Field/AlternateUrls.php]: You can simplify this as follows:

```php
return $this->storeCodeRenderer ??= $this->getLayout()->createBlock(
    // Etc.
);
```

File: src/Block/Adminhtml/Form/Field/StoreCodeColumn.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            $optionArray[] = [
+                'label' =>
+                    $website->getName() .
+                    ' - ' .
+                    $store->getName(),
+                'value' => $store->getId()
+            ];
+        }
+
+        return $optionArray;
```
Comment (janmartenjongerius) [inline src/Block/Adminhtml/Form/Field/StoreCodeColumn.php]: Could you try this with a `return array_map` ? :angel:

File: src/Type/Category.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        return $result;
+    }
+
+    public function getCurrentCategory(): ?CategoryInterface
+    {
+        if (!$this->currentCategory) {
+            $this->currentCategory = $this->registry->registry('current_category');
+        }
+
+        return $this->currentCategory;
```
Comment (janmartenjongerius) [inline src/Type/Category.php]:
```php
return $this->currentCategory ??= $this->registry->registry('current_category');
```

File: src/Type/Category.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (!$currentCategory || !$currentCategory->getId()) {
+            return [];
+        }
+
+        $currentStore = $this->storeManager->getStore();
+        $result       = [];
+
+        foreach ($this->getMapping() as $item) {
+            try {
```
Comment (janmartenjongerius) [inline src/Type/Category.php:58]: This `try-catch` is pretty large / covers multiple expressions.

You could simplify this by adding the following method:

```php
private function getStoreCategory(int $storeId): CategoryInterface
{
    $currentStore = $this->storeManager->getStore();
    $currentCategory = $this->getCurrentCategory();

    return (
        $currentStore->getId() === $storeId
            ? $currentCategory
            : $this->categoryRepository->get(
                $currentCategory->getId(),
                $storeId
            )
    );
}
```

Then, in this method, your `try-catch` looks like:

```php
foreach ($this->getMapping() as $item) {
    try {
        $result[] = [
            'hreflang' => $item['hreflang'],
            'url' => $this->getStoreCategory($item['store_id'])
        ];
    } catch (NoSuchEntityException $e) {
        continue;
    }
}
```

File: src/Type/CmsPage.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    public function getAlternateUrls(): array
+    {
+        /** @var Store $currentStore */
+        $currentStore = $this->storeManager->getStore();
+        $result       = [];
+
+        foreach ($this->getMapping() as $item) {
+            // Because there is no actual link between CMS pages on different
+            // stores we will only return the current store's URL
+            if ($currentStore->getId() === $item['store_id']) {
```
Comment (janmartenjongerius) [inline src/Type/CmsPage.php]: To reduce depth in loops, I suggest inverting checks combined with a continue. So:

```php
foreach ($this->getMapping() as $item) {
    if ($currentStore->getId() !== $item['store_id']) {
        continue;
    }

    $result[] = [];
}
```

File: src/Type/DefaultType.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\Framework\View\Element\Block\ArgumentInterface;
+use Magento\Store\Model\Store;
+use Magento\Store\Model\StoreManagerInterface;
+
+class DefaultType extends AbstractType implements TypeInterface, ArgumentInterface
+{
+    public function __construct(
```
Comment (janmartenjongerius) [inline src/Type/DefaultType.php]: Superflous constructor override

File: src/Type/Product.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        return $result;
+    }
+
+    private function getCurrentProduct(): ?ProductInterface
+    {
+        if (!$this->currentProduct) {
+            $this->currentProduct = $this->registry->registry('current_product');
+        }
+
+        return $this->currentProduct;
```
Comment (janmartenjongerius) [inline src/Type/Product.php]: See previous comments about using `??=`

File: src/view/frontend/templates/alternate_urls.phtml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+use Elgentos\AlternateUrls\Block\AlternateUrls;
+use Elgentos\AlternateUrls\Type\TypeInterface;
+use Magento\Framework\Escaper;
+
+/** @var AlternateUrls $block */
+/** @var Escaper $escaper */
+
+/** @var TypeInterface $typeInstance */
+$typeInstance = $block->getTypeInstance();
```
Comment (janmartenjongerius) [inline src/view/frontend/templates/alternate_urls.phtml:14]: You are assuming the type of this variable above this line, and then checking against falsyness a couple lines later.

Consider the following:

```php

$typeInstance = $block->getTypeInstance();

if (!$typeInstance instanceof TypeInterface) {
    return;
}

foreach ($typeInstance->getAlternateUrls() as $alternateUrl):
    $hreflang = $escaper->escapeHtmlAttr($alternateUrl['hreflang']);
    $href = $escaper->escapeUrl($alternateUrl['url']);
?>
    <link rel="alternate" hreflang="<?= $hreflang ?>" href="<?= $href ?>">
<?php endforeach ?>
```

File: tests/Type/ProductTest.php [hunk 12]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+/**
+ * Copyright MediaCT. All rights reserved.
+ * https://www.mediact.nl
+ */
```
Comment (wahidnory) [inline tests/Type/ProductTest.php:6]: Think this one needs to be removed.

**Review by janmartenjongerius — CHANGES_REQUESTED** [review]
I would really like to see a view model for alternate URLs. Something like:

```php
final class AlternateUrl
{
    private string $href;
    private string $lang;

    public function __construct(string $href, string $lang)
    {
        $this->href = $href;
        $this->lang = $lang;
    }

    public function getHref(Escaper $escaper): string
    {
        return $escaper->escapeUrl($this->href);
    }

    public function getHrefLang(Escaper $escaper): string
    {
        return $escaper->escapeHtmlAttr($this->lang);
    }
}
```

Then, in the type interface, add to the type that it returns `AlternateUrl[]`. And in the block you simplify the code AND enforce that data is properly escaped.

File: src/view/frontend/templates/alternate_urls.phtml (hunk 11 above)
Comment (janmartenjongerius) [reply src/view/frontend/templates/alternate_urls.phtml:14]: See also: https://github.com/elgentos/magento2-alternate-urls/pull/1#pullrequestreview-800961656

File: src/Type/Category.php (hunk 7 above)
Comment (ArjenMiedema) [reply src/Type/Category.php:58]: I assume you need `->getUrl()` behind the `getStoreCategory()` call, right?

**Review by janmartenjongerius — APPROVED** [review]

**Pull request merged**